### PR TITLE
Add sanitizer

### PR DESCRIPTION
**Pull request by sizvix**
Add function to sanitize Manifest before send it to the form generation

### CODE REVIEW

**Comment by sizvix** [comment]
[(𝐒) [APR] Nettoyage du manifeste](https://trello.com/c/kahWjNS0/547-%F0%9D%90%92-apr-nettoyage-du-manifeste)

**Comment by Crash--** [comment]
Tu devrais linter avec nos configs. (Et configurer ton vim pour...) 

J’imagine que le build foire a cause de ça.

File: packages/cozy-harvest/src/Manifest.spec.js [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+      const input1 = {
+        fields: {
+          password: { type: 'password' },
+          plop: { type: 'text' }
+        }
+      }
+      const result = Manifest.sanitizeIdentifier(input1)
+      expect(result.fields.plop.role).toBe('identifier')
+    })
+
```
Comment (gregorylegarec) [inline packages/cozy-harvest/src/Manifest.spec.js]: Another tests to insert here : 

```js
it('should set first non-password field as identifier', () => {
  const manifest = {
    fields: {
      username: {
          type: "text"
       },
      mail: {
         type: 'email'
      },
      password: {
          type: 'password'
      }
  }

  const result = Manifest.sanitize(manifest)
  expect(result.fields.username.role).toBe('identifier')
})

it('should set only one identifier', () => {
  const manifest = {
    fields: {
      username: {
          type: "text"
       },
      mail: {
         type: 'email'
      },
      password: {
         type: 'password'
      }
    }

    const result = Manifest.sanitize(manifest)
    const identifiers = Object.keys(result.fields).filter(fieldName => result.fields[fieldName] === 'identifier')
    expect(identifiers.length).toBe(1)
})

it('should keep identifier priority', () => {
  const manifest = {
    fields: {
      identifier: {
          type: "text"
       },
      username: {
         type: 'email'
      },
      login: {
         type: 'email'
      },
     password: {
         type: 'password'
      }
    }

    const result = Manifest.sanitize(manifest)
    const identifiers = Object.keys(result.fields).filter(fieldName => result.fields[fieldName] === 'identifier')
    expect(identifiers.length).toBe(1)
    expect(result.fields.login.role).toBe('identifier')
})
```

File: packages/cozy-harvest/src/Manifest.spec.js [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  describe('sanitizeIdentifier', () => {
+    it('should set role=identifier for login', () => {
+      const input1 = {
+        fields: {
+          login: { type: 'text' },
+          password: { type: 'password' }
+        }
+      }
+      const result = Manifest.sanitizeIdentifier(input1)
```
Comment (gregorylegarec) [inline packages/cozy-harvest/src/Manifest.spec.js]: I think we should test only the whole sanitize method, and not the internal functions. By doing so we ensure that the sequence of call will keep the change and that another function will not overwrite some properties.

**Comment by gregorylegarec** [comment]
Before merging, the first commit needs to be cleaned:

`git rebase -i HEAD^^`

edit `firstsha1`
pick `secondsha1`

remove merge syntax and then
`git add package.json`
`git rebase --continue`

fix conflicts and then
`git add package.json`
`git rebase --continue`

File: packages/cozy-harvest/src/Manifest.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    if (!manifest.fields) return manifest
+    manifest = Manifest.sanitizeIdentifier(manifest)
+    manifest = Manifest.sanitizeIsRequired(manifest)
+    manifest = Manifest.sanitizeEncrypted(manifest)
+    return manifest
+  },
+
+  sanitizeIdentifier: manifest => {
+    for (let fieldName in manifest.fields)
+      if (manifest.fields[fieldName].role === 'identifier') return manifest
```
Comment (gregorylegarec) [inline packages/cozy-harvest/src/Manifest.js]: I think it does not clean the case when several fields are marked as `identifier`.

File: packages/cozy-harvest/src/Manifest.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+    for (let name of legacyLoginFields)
+      if (manifest.fields[name]) {
+        manifest.fields[name].role = 'identifier'
```
Comment (gregorylegarec) [inline packages/cozy-harvest/src/Manifest.js]: Same, it does not clean several fields marked as `identifiers`

This should need a test imho.

File: packages/cozy-harvest/src/Manifest.js [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        return manifest
+      }
+
+    for (let fieldName in manifest.fields)
+      if (fieldName !== 'password')
+        manifest.fields[fieldName].role = 'identifier'
```
Comment (gregorylegarec) [inline packages/cozy-harvest/src/Manifest.js]: If several fields are not passwords, all will be marked as `identifiers`.

This should need and additional test imho.

File: packages/cozy-harvest/src/Manifest.js [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+    return manifest
+  },
+
+  sanitizeIsRequired: manifest => {
```
Comment (gregorylegarec) [inline packages/cozy-harvest/src/Manifest.js]: Nit : s/sanitizeIsRequired/sanitizeRequired/

File: packages/cozy-harvest/src/Manifest.js [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    for (let fieldName in manifest.fields)
+      if (typeof manifest.fields[fieldName].required != 'boolean')
```
Comment (gregorylegarec) [inline packages/cozy-harvest/src/Manifest.js]: I am surprised that the linter let us used not strict eveluation here. We shoud prefer `!==` over `!=`.

File: packages/cozy-harvest/src/Manifest.spec.js [hunk 8]
```diff
@@ -0,0 +1,136 @@
+/* eslint-env jest */
+import { Manifest } from './Manifest'
+
+describe('Manifest', () => {
+  describe('sanitize', () => {
+    it('should remove "fields" if fields is null', () => {
+      const manifest = {
+        fields: null
```
Comment (gregorylegarec) [inline packages/cozy-harvest/src/Manifest.spec.js]: What happen when manifest is `undefined`? I guess we should have the case of an undeclared `fields` property more often than a `fields` property having a `null` value, and the test should be better with `const manifest = {}`. What do you think ?

My bad : the test is right after 👍

File: packages/cozy-harvest/src/Manifest.spec.js [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+        fields: {
+          password: { type: 'password' },
+          plop: { type: 'text' }
+        }
+      }
+      inputLegacy.fields[name] = { type: 'text' }
+      it('should set role=identifier to ' + name, () => {
+        const result = Manifest.sanitize(inputLegacy)
+        expect(result.fields[name].role).toBe('identifier')
+      })
```
Comment (gregorylegarec) [inline packages/cozy-harvest/src/Manifest.spec.js]:
```js
//...
expect(result.fields.plop.role).not.toBe('identifier')
//...
```

File: packages/cozy-harvest/src/Manifest.spec.js [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+        }
+      }
+      const result = Manifest.sanitize(manifest)
+      const identifiers = Object.keys(result.fields).filter(
+        name => result.fields[name].role === 'identifier'
+      )
+      expect(identifiers.length).toBe(1)
+    })
+
+    it('should keep the identifier pioripy', () => {
```
Comment (gregorylegarec) [inline packages/cozy-harvest/src/Manifest.spec.js]: typo s/prioripy/priority/

File: packages/cozy-harvest/src/Manifest.js [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+  'answer',
+  'access_token',
+  'refresh_token',
+  'appSecret'
+]
+
+export const Manifest = {
+  sanitize: manifest => {
+    if (!manifest.fields) return manifest
+    manifest = Manifest.sanitizeIdentifier(manifest)
```
Comment (gregorylegarec) [inline packages/cozy-harvest/src/Manifest.js]: As a good practice we should not mutate the manifest passed as argument and clone it instead : 
```js
const sanitized = { ...manifest }
sanitized.fields = sanitizedFields(sanitized.fields)
// ...

const sanitizedFields = fields => {
  const sanitized = { ...fields }
  sanitized = sanitizeIdentifier(sanitized)
  return sanitized
}

// etc...
```

With this pattern we can prevent any mutation on the original manifest parameter.

**Comment by gregorylegarec** [comment]
Thanks @sizvix ! 🐻

File: packages/cozy-harvest/src/Manifest.js [hunk 12]
```diff
@@ -17,59 +59,39 @@ export const legacyEncryptedFields = [
...
+ * @param  {Object} [fields={}] Manifest fields
+ * @return {Object}             Sanitized Manifest fields
+ */
+const sanitizeEncrypted = fields => {
+  const sanitized = _cloneDeep(fields)
+  for (let fieldName in sanitized)
+    sanitized[fieldName].encrypted =
+      typeof sanitized[fieldName].encrypted === 'undefined'
+        ? legacyEncryptedFields.includes(fieldName)
+        : !!sanitized[fieldName].encrypted
```
Comment (sizvix) [inline packages/cozy-harvest/src/Manifest.js]: Could be most readable :
```
if (typeof sanitized.fields[fieldName].required !== 'boolean')
  sanitized[fieldName].encrypted = legacyEncryptedFields.includes(fieldName)
```

**Pull request merged**